src/main.rs

```diff
@@ -79,17 +123,108 @@ fn exit_with_error(err: &Error) -> ! {
 
 /// Print error objects in a format intended for end users
 fn print_error(writer: &mut dyn io::Write, err: &Error) -> Result<(), Error> {
-    let mut causes = err.iter_chain();
+    writeln!(writer, "error: {}", err.as_fail())?;
+    for cause in err.iter_causes() {
+        writeln!(writer, "caused by: {}", cause)?;
+    }
+    Ok(())
+}
 
-    if let Some(err_msg) = causes.next() {
-        writeln!(writer, "Error: {}", err_msg)?;
-    } else {
-        writeln!(writer, "Unknown Error")?;
-    };
+fn run_lexer_test<L, O>(lexer: L, out: &mut O) -> Result<(), Error>
+where
+    L: Iterator<Item = TokenKind>,
+    O: io::Write,
+{
+    let token_datas = lexer.filter(|token_data| match token_data {
+        TokenKind::Whitespace | TokenKind::Comment(_) => false,
+        _ => true,
+    });
 
-    for cause in causes {
-        writeln!(writer, "    caused by: {}", cause)?;
+    for td in token_datas {
+        writeln!(out, "{}", td)?;
     }
 
     Ok(())
 }
+
+#[cfg(test)]
+mod lexertest_tests {
+
+    macro_rules! lexer_test_with_tokens {
```

Comment:
As @hediet  pointed out on slack, this could also be a function, instead of a macro. This macro does nothing, a function couldn't do. (A function would even be more concise)